# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/clients/manager/ClientSource.cpp
# Commit: 244df402d6eb
# Full Hash: 244df402d6ebc15d6845c4144245511021408835
# Author: Andrew Sutherland <asutherland@asutherland.org>
# Date: 2024-10-24 09:44:34
# Regressor Bug: 1131324
# File Overlap Count: 1
# Description:
#   Bug 1131324 - Expose ServiceWorkerContainer on WorkerNavigator. r=dom-worker-reviewers,webidl,smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D213725
# ==============================================================================

diff -r 57f70ba5f27e -r 244df402d6eb dom/clients/manager/ClientSource.cpp
--- a/dom/clients/manager/ClientSource.cpp	Thu Oct 24 03:02:39 2024 +0000
+++ b/dom/clients/manager/ClientSource.cpp	Thu Oct 24 03:02:40 2024 +0000
@@ -575,12 +575,12 @@
     const ClientPostMessageArgs& aArgs) {
   NS_ASSERT_OWNINGTHREAD(ClientSource);
 
-  // TODO: Currently this function only supports clients whose global
-  // object is a Window; it should also support those whose global
-  // object is a WorkerGlobalScope.
-  if (nsPIDOMWindowInner* const window = GetInnerWindow()) {
+  // (This only returns a global for windows or workers.)
+  nsIGlobalObject* global = GetGlobal();
+  if (global) {
     const RefPtr<ServiceWorkerContainer> container =
-        window->Navigator()->ServiceWorker();
+        global->GetServiceWorkerContainer();
+    MOZ_ASSERT_DEBUG_OR_FUZZING(container);
 
     // Note, EvictFromBFCache() may delete the ClientSource object
     // when bfcache lives in the child process.
@@ -590,8 +590,7 @@
   }
 
   CopyableErrorResult rv;
-  rv.ThrowNotSupportedError(
-      "postMessage to non-Window clients is not supported yet");
+  rv.ThrowInvalidStateError("Global discarded");
   return ClientOpPromise::CreateAndReject(rv, __func__);
 }
 